tests/components/metoffice/test_metoffice.py

```diff
@@ -0,0 +1,327 @@
+    "feels_like_temperature": ("feels_like_temperature", "13"),
+    "uv": ("uv_index", "6"),
+    "precipitation": ("probability_of_precipitation", "0"),
+    "wind_direction": ("wind_direction", "E"),
+    "wind_gust": ("wind_gust", "7"),
+    "wind_speed": ("wind_speed", "2"),
+    "humidity": ("humidity", "60"),
+}
+
+WAVERTREE_SENSOR_RESULTS = {
+    "weather": ("weather", "sunny"),
+    "visibility": ("visibility", "Good"),
+    "visibility_distance": ("visibility_distance", "10-20"),
+    "temperature": ("temperature", "17"),
+    "feels_like_temperature": ("feels_like_temperature", "14"),
+    "uv": ("uv_index", "5"),
+    "precipitation": ("probability_of_precipitation", "0"),
+    "wind_direction": ("wind_direction", "SSE"),
+    "wind_gust": ("wind_gust", "16"),
+    "wind_speed": ("wind_speed", "9"),
+    "humidity": ("humidity", "50"),
+}
+
+
+class MockDateTime(datetime):
+    """Replacement for datetime that can be mocked for testing."""
+
+    def __new__(cls, *args, **kwargs):
+        """Override to just return base class."""
+        return datetime.__new__(datetime, *args, **kwargs)
+
+
+@patch("datetime.datetime", MockDateTime)
+async def test_one_weather_site_running(hass, requests_mock):
+    """Test the Met Office weather platform."""
+
+    # all metoffice test data encapsulated in here
+    mock_json = json.loads(load_fixture("metoffice.json"))
+    all_sites = json.dumps(mock_json["all_sites"])
+    wavertree_hourly = json.dumps(mock_json["wavertree_hourly"])
+
+    from datetime import datetime, timezone
+
+    MockDateTime.now = classmethod(
+        lambda *args, **kwargs: datetime(2020, 4, 25, 12, tzinfo=timezone.utc)
+    )
+
+    requests_mock.get("/public/data/val/wxfcs/all/json/sitelist/", text=all_sites)
+    requests_mock.get(
+        "/public/data/val/wxfcs/all/json/354107?res=3hourly", text=wavertree_hourly,
+    )
+
+    entry = MockConfigEntry(domain=DOMAIN, data=METOFFICE_CONFIG_WAVERTREE,)
+    entry.add_to_hass(hass)
+    await hass.config_entries.async_setup(entry.entry_id)
+    await hass.async_block_till_done()
+
+    # Wavertree weather platform expected results
+    entity = hass.states.get("weather.met_office_wavertree")
+    assert entity
+
+    assert entity.state == "sunny"
+    assert entity.attributes.get("temperature") == 17
+    assert entity.attributes.get("wind_speed") == 9
+    assert entity.attributes.get("wind_bearing") == "SSE"
+    assert entity.attributes.get("visibility") == "Good - 10-20"
+    assert entity.attributes.get("humidity") == 50
+
+
+@patch("datetime.datetime", MockDateTime)
+async def test_one_sensor_site_running(hass, requests_mock):
+    """Test the Met Office sensor platform."""
+
+    # all metoffice test data encapsulated in here
+    mock_json = json.loads(load_fixture("metoffice.json"))
+    all_sites = json.dumps(mock_json["all_sites"])
+    wavertree_hourly = json.dumps(mock_json["wavertree_hourly"])
+
+    from datetime import datetime, timezone
```

Comment:
Standard library imports should be at the top of the module.

Reply:
They should, yes, but the patch to override "now" doesn't work unless you import datetime at the point you need to mess with it (as per my searching which led to [this document](http://williambert.online/2011/07/how-to-unit-testing-in-django-with-mocking-and-patching/)).

Reply:
Where is `datetime.datetime` used? This might be a matter of patching in the correct place:
https://docs.python.org/3/library/unittest.mock.html#where-to-patch

Reply:
It's used as the datapoint library checks the date/time and returns the "current" data for the forecast site. I've mocked the API call to feed it some specific data for a given date/time but then need the API to pass me predictable data to test against. Hence the need to mock the date/time.

Reply:
Have you tried patching datetime where it's used?

Reply:
I did try, but maybe it's my lack of experience with Python mocks, I couldn't get it to work. This was the only way I could seem to get it to behave in a consistent way and mock "now" so that datapoint API returned the expected data.